cardano-api/src/Cardano/Api/Fees.hs

```diff
@@ -766,72 +793,141 @@ makeTransactionBodyAutoBalance eraInMode systemstart history pparams
                        | otherwise         -> Left (TxBodyScriptExecutionError
                                                      (Map.toList failures))
 
-    let txbodycontent1 = substituteExecutionUnits exUnitsMap' txbodycontent
+    let txbodycontentWithExecUnits = substituteExecutionUnits exUnitsMap' txbodycontent
 
     explicitTxFees <- first (const TxBodyErrorByronEraNotSupported) $
-                        txFeesExplicitInEra era'
+                        txFeesExplicitInEra (shelleyBasedToCardanoEra sbe)
 
     -- Insert change address and set tx fee to 0
-    txbody1 <- first TxBodyError $ -- TODO: impossible to fail now
-               makeTransactionBody txbodycontent1 {
-                 txFee  = TxFeeExplicit explicitTxFees 0,
-                 txOuts = TxOut changeaddr
-                                (lovelaceToTxOutValue 0)
-                                TxOutDatumHashNone
-                        : txOuts txbodycontent
-                 --TODO: think about the size of the change output
-                 -- 1,2,4 or 8 bytes?
-               }
+    txbodyZeroFeesChangeOutputZero <-
+      first TxBodyError $ -- TODO: impossible to fail now
+        obtainIsEra sbe $ makeTransactionBody txbodycontentWithExecUnits
+          { txFee  = TxFeeExplicit explicitTxFees 0 }
 
-    let nkeys = fromMaybe (estimateTransactionKeyWitnessCount txbodycontent1)
+    eUnitsCost <- sumExecUnits exUnitsMap'
+    let nkeys = fromMaybe (estimateTransactionKeyWitnessCount txbodycontentWithExecUnits)
                           mnkeys
-        fee   = evaluateTransactionFee pparams txbody1 nkeys 0 --TODO: byron keys
+        fee   = obtainIsCardanoEraConstraint sbe $ evaluateTransactionFee pparams txbodyZeroFeesChangeOutputZero nkeys 0 --TODO: byron keys
+        totalFee = eUnitsCost + fee
```

Comment:
I don't understand this bit. Why do we have to separately calculate the ex unit costs and add that to the fee. The `evaluateTransactionFee` is supposed to calculate the fee for us, including any ex units. That's what `evaluateTransactionFee` is supposed to be there for (since just size based is easy and we didn't need ledger lib support for that). Sounds like the ledger function is not doing what we want and we're having to work around it here.